core/src/main/java/io/questdb/cairo/sql/async/PageFrameSequence.java

```diff
@@ -0,0 +1,440 @@
+    private final LongList frameRowCounts = new LongList();
+    private final PageFrameReducer reducer;
+    private final PageAddressCache pageAddressCache;
+    private final MessageBus messageBus;
+    private final AtomicInteger owner = new AtomicInteger(OWNER_NONE);
+    private final MicrosecondClock microsecondClock;
+    private long id;
+    private int shard;
+    private int frameCount;
+    private Sequence collectSubSeq;
+    private SymbolTableSource symbolTableSource;
+    private T atom;
+    private PageAddressCacheRecord[] records;
+    // we need this to restart execution for `toTop`
+    private MPSequence dispatchPubSeq;
+    private RingQueue<PageFrameDispatchTask> pageFrameDispatchQueue;
+    private final AtomicInteger dispatchStartIndex = new AtomicInteger();
+    private SqlExecutionCircuitBreaker[] circuitBreakers;
+    private long startTimeUs;
+    private long circuitBreakerFd;
+
+    public PageFrameSequence(
+            CairoConfiguration configuration,
+            MessageBus messageBus,
+            PageFrameReducer reducer
+    ) {
+        this.pageAddressCache = new PageAddressCache(configuration);
+        this.messageBus = messageBus;
+        this.reducer = reducer;
+        this.microsecondClock = configuration.getMicrosecondClock();
+    }
+
+    /**
+     * Waits for frame sequence completion, fetches remaining pieces of the
+     * frame sequence from the queues. This method is not thread safe.
+     */
+    public void await() {
+        LOG.debug()
+                .$("awaiting completion [shard=").$(shard)
+                .$(", id=").$(id)
+                .$(", frameCount=").$(frameCount)
+                .I$();
+
+        // Collect only dispatched frames. To do that we need to stall any further dispatch.
+        while (true) {
+            final int dispatched = dispatchStartIndex.get();
+            if (dispatchStartIndex.compareAndSet(dispatched, frameCount)) {
+                frameCount = dispatched;
+                break;
+            }
+        }
+
+        final RingQueue<PageFrameReduceTask> queue = getPageFrameReduceQueue();
+        final MCSequence pageFrameReduceSubSeq = messageBus.getPageFrameReduceSubSeq(shard);
+        while (doneLatch.getCount() == 0) {
+            final int workerId = getWorkerId();
+            final PageAddressCacheRecord rec = records[workerId];
+            final SqlExecutionCircuitBreaker circuitBreaker = circuitBreakers[getWorkerId()];
+            // we were asked to steal work from dispatch0 queue and beyond, as much as we can
+            if (PageFrameReduceJob.consumeQueue(queue, pageFrameReduceSubSeq, rec, circuitBreaker)) {
+                long cursor = collectSubSeq.next();
+                if (cursor > -1) {
+                    // discard collect items
+                    final PageFrameReduceTask tsk = queue.get(cursor);
+                    if (tsk.getFrameSequence() == this) {
+                        tsk.collected();
```

Comment:
nit: might be a good idea to introduce an input parameter in `collected()` method to force `resetCapacity()` call on the row id list. This would free some memory in early exit scenarios (`close()`/`toTop()`/etc.).

Reply:
This is by design. Forcing reset by collecting code would be without context or any extra context that `collect()` is already aware of. Resetting list incorrectly will cause extra reallocations.

Reply:
both `close()` and `top()` have to drain the queue regardless and they will reset the rowid lists

Reply:
Makes sense, thanks for the clarification.